ribs.py

```diff
@@ -0,0 +1,132 @@
+#
+# The aim of this file is to give a collection of usable functions
+# and structures to be used as a jumping off point. Feel free
+# to change and distribute!
+
+import pygame as pg
+# For smaller cheat sheet see:
+#   https://www.lithekod.se/pygame.html
+# For official pygame documentation see:
+#   https://www.pygame.org/docs/
+
+#
+# Utility
+#
+
+running = True
+def is_running():
+    """Is the game currently running?"""
+    return running
+
+
+def quit_game():
+    """Pull the brakes and stop the game!"""
+    global running
+    running = False
+
+
+#
+# Input handling
+#
+
+def event_is(event, kind):
+    return pg.event.event_name(event.type) == kind
+
+
+current_frame_held_buttons = set()
+last_frame_held_buttons = None
+def process_events():
+    """Tells the game what buttons are pressed."""
+    global last_frame_held_buttons
+    last_frame_held_buttons = current_frame_held_buttons.copy()
+    for event in pg.event.get():
+        if event_is(event, "Quit"):
+            quit_game()
+        elif event_is(event, "KeyDown"):
+            current_frame_held_buttons.add(event.key)
+        elif event_is(event, "KeyUp"):
+            current_frame_held_buttons.remove(event.key)
+        elif event_is(event, "MouseButtonUp"):
+            ...
+        elif event_is(event, "MouseButtonDown"):
+            ...
+        elif event_is(event, "MouseMotion"):
+            ...
+
+
+def to_keycode(key):
+    """Takes a keycode or a character and converts it to a keycode"""
+    if type(key) == str:
+        if len(key) != 1:
+            raise "error"
+        return ord(key.lower())
+    return key
+
+
+def key_down(key):
+    """
+        Takes a key, that's either a keycode or a character,
+        and says if it's down or not
+    """
+    keycode = to_keycode(key)
+    return keycode in current_frame_held_buttons
+
+
+def key_released(key):
+    """
+        Takes a key, that's either a keycode or a character,
+        and says if it was released this frame.
+    """
+    keycode = to_keycode(key)
+    return (keycode not in current_frame_held_buttons) and \
+           (keycode in last_frame_held_buttons)
+
+
+def key_pressed(key):
+    """
+        Takes a key, that's either a keycode or a character,
+        and says if it was pressed down this frame.
+    """
+    keycode = to_keycode(key)
+    return (keycode in current_frame_held_buttons) and \
+           (keycode not in last_frame_held_buttons)
+
+
+#
+# Main loop
+#
+
+SCREEN_WIDTH = 500
+SCREEN_HEIGHT = 500
+def main():
+    """The program starts here"""
+    pg.init()
+    pg.display.init()
+
+    # Sets the screen resolution.
+    pg.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
+
+    frame_clock = pg.time.Clock()
+
+    while is_running():
+        # Tell Pygame we're on a new frame, with the given framerate
+        # set it to zero to unlimit.
+        frame_clock.tick(60)
+        # See what buttons are pressed this frame.
+        process_events()
```

Comment:
Maybe we can get rid of the global here by making process_events take the previous events as an argument and returning a new block.

Reply:
We could do this, but then you only have one way to stop the game. If you want to add a mouse button or something that stops the game, you can't really do that.

I don't know what is the simplest to read and understand. It might be that global state makes it less readable, or it makes it more confusing since you don't know where stuff happens anyways.

Reply:
Is stopping the game really a central thing people are going to want to do?

Also, my objection is primarily based on not teaching new people bad habbits. This might be the first "large" code base they see, and we should avoid obvious anti-patterns like globals then in my opinion